maths/cholesky_decomposition.py

```diff
@@ -0,0 +1,102 @@
+import numpy as np
+
+
+def cholesky_decomposition(a: np.ndarray) -> np.ndarray:
+    """Return a Cholesky decomposition of the matrix A.
+
+    The Cholesky decomposition decomposes the square, positive definite matrix A
+    into a lower triangular matrix L such that A = L L^T.
+
+    https://en.wikipedia.org/wiki/Cholesky_decomposition
+
+    Arguments:
+    A -- a numpy.ndarray of shape (n, n)
+
+    >>> A = np.array([[4, 12, -16], [12, 37, -43], [-16, -43, 98]], dtype=float)
+    >>> L = cholesky_decomposition(A)
+    >>> np.allclose(L, np.array([[2, 0, 0], [6, 1, 0], [-8, 5, 3]]))
+    True
+
+    >>> # check that the decomposition is correct
+    >>> np.allclose(L @ L.T, A)
+    True
+
+    >>> # check that L is lower triangular
+    >>> np.allclose(np.tril(L), L)
+    True
+
+    The Cholesky decomposition can be used to solve the system of equations A x = y.
+
+    >>> x_true = np.array([1, 2, 3], dtype=float)
+    >>> y = A @ x_true
+    >>> x = solve_cholesky(L, y)
+    >>> np.allclose(x, x_true)
+    True
+
+    It can also be used to solve multiple equations A X = Y simultaneously.
+
+    >>> X_true = np.random.rand(3, 3)
+    >>> Y = A @ X_true
+    >>> X = solve_cholesky(L, Y)
+    >>> np.allclose(X, X_true)
+    True
+    """
+    assert a.shape[0] == a.shape[1]
+    n = a.shape[0]
+    lo = np.tril(a)
```

Comment:
can you give the `lo` variable a proper name like `lower_triangle`?

Reply:
I renamed `lo` to `L`. Now the code lines up nicely with the variable naming scheme on the Wikipedia article about the Cholesky decomposition, which makes it easier to understand. SciPy also use upper case single letter variable names for matrices in some decomposition algorithms, so there definitely is a precedence for that https://docs.scipy.org/doc/scipy/reference/linalg.html#decompositions